### ISSUE TITLE
Implement JsonRpcHttpConsumer?

### ISSUE THREAD

**Issue by ekzobrain** [millerf/django-channels-jsonrpc#8]
Hi.

What about implementing another consumer, for http requests in the same fashion as websocket?

**Comment by millerf** [comment]
I developed this one for one of my specific project, I guess it shouldn't be that difficult to extend to Channels' [BaseConsumer](https://channels.readthedocs.io/en/stable/generics.html#base).
You are more than welcome to extend what has been done and submit a PR if you want...

**Comment by ekzobrain** [comment]
I see that you also do not have inbound notifications support...
Do you plan to implement it anytime soon? or do you have any plans on how it might work?

**Comment by millerf** [comment]
I will definitely implement it soon.  Again, feel free to start a branch...
No plans just yet, no.

**Comment by millerf** [comment]
Also, please open another issue, as the 2 questions are not correlated.

**Comment by ekzobrain** [comment]
Http implementation needs to have just one method for processing HTTP request and the same decoding/encoding logic.
I think to implement HttpConsumer current JsonRpcWebsocketConsumer may be just renamed to JsonRpcConsumer (also more straight-forward with package name) and a class method like "http_handler()" may be added to handle http requests.
Then one could point to it this way:
```python
channel_routing = [
    JsonRpcConsumer.as_route(path=r'/ws')
    route('http.request', JsonRpcConsumer.http_handler, path=r'/http')
]
```
I think it would be the most reliable way. Registered rpc_method handlers then could be called both via HTTP and WebSocket.

**Comment by millerf** [comment]
Yes that sounds like a good implementation plan.
What about notifications, though? I still don't understand the need of an implementation of JSON-RPC over HTTP and how it can be fully implemented...
There is a project implementing that (but outside of Channels) here:
https://github.com/samuraisam/django-json-rpc
Maybe there is some idea that can be reused? I haven't went through it just yet...

**Comment by ekzobrain** [comment]
Yes, I am currently looking into this implementation. It is a simple logic over Django's view layer. And it could be used as a fallback, but I do not want to spread my handlers and processing logic to different places.
In my current project I have several services/interfaces talking to my Django-Channels backend via JSON-RPC. But transfers may differ: Browsers and Mobile Apps connect via WebSocket, and some Software backend services (telephony stations in my case) are only capable of JSON-RPC via HTTP. So I need to handle both WebSocket and HTTP based requests.
Of course HTTP implementation could support only inbound notifications. HTTP-based JSON-RPC calls are one-directional. For outbound requests in this case special Client library must be used which wraps CURL requests, for example.

**Comment by millerf** [comment]
Yep I do understand your point-of-view.
About outbound notifications, you might be able to do it by having a "pool" of notifications buffered on the server, and polling every X seconds to retrieve all of them. Not perfect, but a nice fallback. That could be done on the application layer, but it could also be done in this module for the application to use it...
Not sure I am still making sense. It is starting to get late here :)

**Comment by ekzobrain** [comment]
Channels support HTTP2 sessions, but I did not look into how to use it. I think outbound notifications may be supported in this way via HTTP2 PUSH.

**Comment by millerf** [comment]
I was talking about mere HTTP. I am sure your "Software backend services" do not support don't support HTTP/2, are they?

**Comment by ekzobrain** [comment]
Till we already have wide HTTP2 support, implementing a hack via HTTP1.1 would not be much usefull. HTTP1.1 was not created for bi-directional communications.

**Comment by ekzobrain** [comment]
Yes, they do not, currently.

**Comment by millerf** [comment]
I was trying to create a fallback for HTTP1, in order to keep the same features (evens somewhat degraded) across all protocols

**Comment by ekzobrain** [comment]
Of course this may be done, but till this is an extension for Channels, it should serve the same purposes. And send outbound HTTP/1 (JSON-RPC encoded) requests is not in Channels functional responsibility. Channels is for serving persistent requests (WebSockets,HTTP/2), so this sould be fully implemented first. Ability to serve HTTP/1 inbound calls is just a "handy entension", which allows not using extra JSON-RPC Server package and maintain code cleanless. For outbound JSON-RPC HTTP/1 calls it is always OK to use any python JSON-RPC Client implementation.

**Comment by millerf** [comment]
Ok, I get you now.

**Comment by ekzobrain** [comment]
Hi.

I've implemented first version of HTTP/1 support. You may review it here: https://github.com/soulhunter1987/django-channels-jsonrpc/tree/add_http_handler
It is not tested, I just wan't you to take a look over it and provide some feedback, then I will test it and create a pull request.
I also removed threads there and made some minor cleanups.

**Comment by millerf** [comment]
Nice.

Thanks for the - multiple - spelling mistakes :)

Removing the threading will also greatly help with retrieving the "original_message", which has been a pain to transmit to the RPC methods. I'll have a look at that after I merge it.
Could you also please try and test that the method calls are asynchronous. I am looking into it and it doesn't seem channels is running all the consumers asyncronously...

From http://channels.readthedocs.io/en/stable/inshort.html#do-i-need-to-worry-about-making-all-my-code-async-friendly
> > Do I need to worry about making all my code async-friendly?
> No, all your code runs synchronously without any sockets or event loops to block. You can use async code within a Django view or channel consumer if you like - for example, to fetch lots of URLs in parallel - but it doesn’t affect the overall deployed site.
> 


But thanks anyway for the nice work!

**Comment by ekzobrain** [comment]
Yes, I have some more ideas to implement:
1. Resticting rpc methods to some transport. By default a consumer can handle both websocket and http transports with the same handlers, but it may be configured per method, for example:
This method will be available only through websocket:
```python
@MyJsonRpcConsumer.rpc_method(http=False)
def handler():
    return ''
```
And whis one only through http:
```python
@MyJsonRpcConsumer.rpc_method(websocket=False)
def handler():
    return ''
```
This implies this routing, meaning all possible requests (websocket.connect, websocket.message, websocket.disconnect, http.request) will be handled by this consumer:
```python
channel_routing = [
    MyJsonRpcConsumer.as_route()
]
```

Of couse, user may restrict transport on routing level. With this configuration MyJsonRpcConsumer will never recieve websocket events, so it is guaranteed to handle only http requests:
```python
channel_routing = [
    route_class('http.request', MyJsonRpcConsumer)
]
```

2. Authentication - any method may require or not user authentication. This method won't be available for anonymous users. We just need to decide what result to provide then:
```python
@MyJsonRpcConsumer.rpc_method(authenticated=True)
def handler():
    return ''
```

**Comment by millerf** [comment]
Yes, sounds perfect.

**Comment by ekzobrain** [comment]
Hi.

Implementation is tested and ready to merge. I've implemented per-method protocol restrictions, but not authentication checks, I'll do it a little bit later.

**Comment by ekzobrain** [comment]
I also changed a way "original_message" is passed to rpc methods. I think to add it as an explicit extra parameter to method handlers is not an optimal solution. It is better to use whis syntax:

```python
def hander(param1, param2, **kwargs):
    message = kwargs['original_message']
    pass
```
It is closer to Channels way of passing aditional "environment" params to callbacks and allows to visually distinguish method parameters and extra "environment" parameters.

**Comment by ekzobrain** [comment]
Would you, please, create a branch called "add_http_handler", so that I could make a pull request to it?

**Comment by millerf** [comment]
Ok, agreed. 
Branch has been created

**Comment by millerf** [comment]
I don't see much in the README about the Http/1 part...
Also, did you force a format in the rpc_method names, [here](https://github.com/millerf/django-channels-jsonrpc/blob/add_http_handler/channels_jsonrpc/jsonrpcconsumer.py#L259)?

**Comment by millerf** [comment]
nevermind :)

**Comment by ekzobrain** [comment]
I will add more info to README soon

**Comment by millerf** [comment]
```
def hander(param1, param2, **kwargs):
    message = kwargs['original_message']
    pass
```

I am fighting with this, as if you NEED to have at least **kwargs if the method takes no argument... No sure if that is the best way...

**Comment by millerf** [comment]
@soulhunter1987  I made the PR, please have a look at it and also add documentation. 
Thanks. I am still not 100% sure for the way to handle `original_message`...
